src/main/java/com/hubspot/jinjava/util/DeferredValueUtils.java

```diff
@@ -89,13 +89,20 @@ public static void deferVariables(String[] varTokens, Map<String, Object> contex
     Set<String> deferredProps = getPropertiesUsedInDeferredNodes(context, templateSource);
     Set<String> setProps = getPropertiesSetInDeferredNodes(templateSource);
     if (eagerToken != null) {
-      deferredProps.addAll(
-        getPropertiesUsedInDeferredNodes(
-          context,
-          rebuildTemplateForEagerTagTokens(eagerToken, true),
-          false
-        )
-      );
+      if (
+        eagerToken.getCurrentMacroFunction() == null ||
+        eagerToken
+          .getCurrentMacroFunction()
+          .equals(context.getMacroStack().peek().orElse(null))
+      ) {
+        deferredProps.addAll(
+          getPropertiesUsedInDeferredNodes(
+            context,
+            rebuildTemplateForEagerTagTokens(eagerToken, true),
+            false
+          )
+        );
```

Comment:
We only run this check for the variables that are deferred from being `set` because those are the ones that are attempted to be `put` onto the context. Therefore, those are the ones whose scope we need to limit. If trying to modify a variable that exists outside the macro function's context through something like `append()` within the macro function, we would still want to defer that as those changes propagate outside of the macro function's scope